File: absurdityAppForJava/src/main/java/forOlderJava/absurdityAppForJava/domain/entity/Order.java
package forOlderJava.absurdityAppForJava.domain.entity;

import forOlderJava.absurdityAppForJava.global.exception.CustomException;
import forOlderJava.absurdityAppForJava.global.exception.ErrorCode;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.List;

@Entity
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Getter
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "ORDER_ID")
    private Long id;

    @Embedded
    private OrderInfo orderInfo;

    @Embedded
    private Orderer orderer;

    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinColumn(name = "ORDER_ID")
    private List<OrderItem> orderItems;

    public void requestCancel() throws CustomException {
        if (orderInfo.getOrderStatus() == OrderStatus.SHIPPING ||
        orderInfo.getOrderStatus() == OrderStatus.COME_OUT ||
        orderInfo.getOrderStatus() == OrderStatus.COOKING) {
            throw new CustomException(ErrorCode.ORDER_NOT_FOUND);
        }
        orderInfo.setOrderStatus(OrderStatus.OLDER_CANCEL_REQUEST);
    }
}

File: ForOlder/src/main/kotlin/absurdity/ForOlder/domain/Order.kt
package absurdity.ForOlder.domain

import absurdity.ForOlder.common.exception.CustomException
import absurdity.ForOlder.common.exception.ErrorCode.CANNOT_CANCEL_ORDER
import absurdity.ForOlder.kafka.dto.OrderItem

data class Order (
    val orderInfo: OrderInfo,
    val orderer: Orderer,
    val orderItem: List<OrderItem>,
) {
    fun requestCancel() {
        // 반항하고싶거나 피치 못할 사정이면 취소가 가능하다.
        if (orderInfo.orderStatus == OrderStatus.SHIPPING || orderInfo.orderStatus == OrderStatus.COOKING || orderInfo.orderStatus == OrderStatus.COME_OUT) {
            throw CustomException(CANNOT_CANCEL_ORDER)
        }
        orderInfo.orderStatus = OrderStatus.OLDER_CANCEL_REQUEST
    }
}
